ReactiveCocoa/Swift/Property.swift

```diff
@@ -8,66 +8,457 @@ public protocol PropertyType {
+public func zip<S: SequenceType where S.Generator.Element: PropertyType>(properties: S) -> AnyProperty<[S.Generator.Element.Value]>? {
+	var generator = properties.generate()
+	if let first = generator.next() {
+		let initial = first.map { [$0] }
+		return GeneratorSequence(generator).reduce(initial) { property, next in
+			property.zip(with: next).map { $0.0 + [$0.1] }
+		}
+	}
+
+	return nil
+}
+
+/// A read-only, type-erased view of a property.
 public struct AnyProperty<Value>: PropertyType {
+	private let capturingClosure: (() -> Void)?
+	private let disposable: Disposable?
 
 	private let _value: () -> Value
 	private let _producer: () -> SignalProducer<Value, NoError>
 	private let _signal: () -> Signal<Value, NoError>
 
-
+	/// The current value of the property.
 	public var value: Value {
 		return _value()
 	}
 
+	/// A producer for Signals that will send the wrapped property's current value,
+	/// followed by all changes over time, then complete when the wrapped property has
+	/// deinitialized.
 	public var producer: SignalProducer<Value, NoError> {
 		return _producer()
 	}
 
+	/// A signal that will send the wrapped property's changes over time, then complete
+	/// when the wrapped property has deinitialized.
+	///
+	/// It is strongly discouraged to use `signal` on any transformed property.
 	public var signal: Signal<Value, NoError> {
 		return _signal()
 	}
-	
+
 	/// Initializes a property as a read-only view of the given property.
 	public init<P: PropertyType where P.Value == Value>(_ property: P) {
+		capturingClosure = AnyProperty.capture(property)
+		disposable = nil
 		_value = { property.value }
 		_producer = { property.producer }
 		_signal = { property.signal }
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on a signal created by `producer`.
 	public init(initialValue: Value, producer: SignalProducer<Value, NoError>) {
 		let mutableProperty = MutableProperty(initialValue)
 		mutableProperty <~ producer
 		self.init(mutableProperty)
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on `signal`.
 	public init(initialValue: Value, signal: Signal<Value, NoError>) {
 		let mutableProperty = MutableProperty(initialValue)
 		mutableProperty <~ signal
 		self.init(mutableProperty)
 	}
-}
 
-extension PropertyType {
-	/// Maps the current value and all subsequent values to a new value.
-	public func map<U>(transform: Value -> U) -> AnyProperty<U> {
-		let mappedProducer = SignalProducer<U, NoError> { observer, disposable in
-			disposable += ActionDisposable { self }
-			disposable += self.producer.map(transform).start(observer)
+	/// Initializes a property by applying the unary `SignalProducer` transform on
+	/// `property`. The resulting property captures `property`.
+	private init<P: PropertyType>(transforming property: P, @noescape using transform: SignalProducer<P.Value, NoError> -> SignalProducer<Value, NoError>) {
```

Comment:
But it looks better at the call site. Instead of `AnyProperty(property, transform: transform)`, we have  `AnyProperty(property, using: transform)` that IMO makes more sense.